File: send_notif.py
"""
*********************************************************************
* YANG Book Send Purchase Notification                              *
* Derived from: ConfD Actions intro example                         *
*                                                                   *
* (C) 2018 Jan Lindblad                                             *
*                                                                   *
* See the LICENSE file for license information                      *
* See the README  file for more information                         *
*********************************************************************
"""
import time
import confd
from confd.dp import Daemon
from confd.maapi import Maapi
import _confd.dp as lowdp
from bookzone_example_ns import ns

def send_shipping_notif(stream, user, title, fmt, copies):
  values = [
    confd.TagValue(confd.XmlTag(ns.hash, ns.bz_shipping), confd.Value((ns.bz_shipping, ns.hash), confd.C_XMLBEGIN)),
    confd.TagValue(confd.XmlTag(ns.hash, ns.bz_user),     confd.Value(user)),
    confd.TagValue(confd.XmlTag(ns.hash, ns.bz_title),    confd.Value(title)),
    confd.TagValue(confd.XmlTag(ns.hash, ns.bz_format),   confd.Value((ns.hash, fmt), confd.C_IDENTITYREF)),
    confd.TagValue(confd.XmlTag(ns.hash, ns.bz_copies),   confd.Value(copies, confd.C_UINT32)),
    confd.TagValue(confd.XmlTag(ns.hash, ns.bz_shipping), confd.Value((ns.bz_shipping, ns.hash), confd.C_XMLEND))
  ]
  gm = time.gmtime(time.time())
  now = confd.DateTime(gm.tm_year, gm.tm_mon, gm.tm_mday, 
    gm.tm_hour, gm.tm_min, gm.tm_sec, 0, 0, 0)
  lowdp.notification_send(stream, now, values)

if __name__ == "__main__":
    try:
      d = Daemon(name='purchase-notification-sender-daemon', log=None)
      workersock = confd.dp.take_worker_socket(d, 'trader', 'send')
      trader = lowdp.register_notification_stream(d.ctx(), None, workersock, 'Trader')
      d.start()
      title = "What We Think About When We Try Not To Think About Global Warming: Toward a New Psychology of Climate Action"
      send_shipping_notif(trader, user="janl", title=title, fmt=ns.bz_paperback, copies=1)
      confd.dp.return_worker_socket(d, 'send')
    finally:
      d.finish()

File: common/send_notif.c
/********************************************************************
* YANG Book Send Purchase Notification                              *
* Derived from: NETCONF Notifications example                       *
*                                                                   *
* (C) 2018 Jan Lindblad                                             *
*                                                                   *
* See the LICENSE file for license information                      *
* See the README  file for more information                         *
********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/poll.h>
#include <time.h>
#include <sys/time.h>

#include <confd_lib.h>
#include <confd_dp.h>

#include "bookzone-example.h"

#define OK(val) (assert((val) == CONFD_OK))
static int ctlsock, workersock;
static struct confd_daemon_ctx *dctx;

/* The notification context (filled in by ConfD) for the live feed */
static struct confd_notification_ctx *live_ctx;

static int get_ctlsock(struct addrinfo *addr)
{
    int sock;

    if ((sock =
         socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol)) < 0)
        return -1;
    if (confd_connect(dctx, sock, CONTROL_SOCKET,
                      addr->ai_addr, addr->ai_addrlen) != CONFD_OK) {
        close(sock);
        return -1;
    }
    return sock;
}

static int get_workersock(struct addrinfo *addr)
{
    int sock;

    if ((sock =
         socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol)) < 0)
        return -1;
    if (confd_connect(dctx, sock, WORKER_SOCKET,
                      addr->ai_addr, addr->ai_addrlen) != CONFD_OK) {
        close(sock);
        return -1;
    }
    return sock;
}

static void getdatetime(struct confd_datetime *datetime)
{
    struct tm tm;
    struct timeval tv;

    gettimeofday(&tv, NULL);
    gmtime_r(&tv.tv_sec, &tm);

    memset(datetime, 0, sizeof(*datetime));
    datetime->year = 1900 + tm.tm_year;
    datetime->month = tm.tm_mon + 1;
    datetime->day = tm.tm_mday;
    datetime->sec = tm.tm_sec;
    datetime->micro = tv.tv_usec;
    datetime->timezone = 0;
    datetime->timezone_minutes = 0;
    datetime->hour = tm.tm_hour;
    datetime->min = tm.tm_min;
}

static void send_shipping_notif(char*user, char*title, struct confd_identityref fmt, int copies)
{
    confd_tag_value_t vals[9];
    struct confd_datetime now;
    int i = 0;

    CONFD_SET_TAG_XMLBEGIN   (&vals[i], bz_shipping,         bz__ns); i++;
    CONFD_SET_TAG_STR        (&vals[i], bz_user,             user);   i++;
    CONFD_SET_TAG_STR        (&vals[i], bz_title,            title);  i++;
    CONFD_SET_TAG_IDENTITYREF(&vals[i], bz_format,           fmt);    i++;
    CONFD_SET_TAG_UINT32     (&vals[i], bz_number_of_copies, copies); i++;
    CONFD_SET_TAG_XMLEND     (&vals[i], bz_shipping,         bz__ns); i++;
    getdatetime(&now);
    OK(confd_notification_send(live_ctx, &now, vals, i));
    fprintf(stderr, "%s\n", "Trader Shipping notification delivered");
}

int main(int argc, char **argv)
{
    char confd_port[16];
    struct addrinfo hints;
    struct addrinfo *addr = NULL;
    int debuglevel = CONFD_SILENT;
    int i;
    int c;
    char *p, *dname;
    struct confd_notification_stream_cbs ncb;

    snprintf(confd_port, sizeof(confd_port), "%d", CONFD_PORT);
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = PF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    while ((c = getopt(argc, argv, "dtprc:")) != -1) {
        switch (c) {
        case 'd':
            debuglevel = CONFD_DEBUG;
            break;
        case 't':
            debuglevel = CONFD_TRACE;
            break;
        case 'p':
            debuglevel = CONFD_PROTO_TRACE;
            break;
        case 'c':
            if ((p = strchr(optarg, '/')) != NULL)
                *p++ = '\0';
            else
                p = confd_port;
            if (getaddrinfo(optarg, p, &hints, &addr) != 0) {
                if (p != confd_port) {
                    *--p = '/';
                    p = "/port";
                } else {
                    p = "";
                }
                fprintf(stderr, "%s: Invalid address%s: %s\n",
                        argv[0], p, optarg);
                exit(1);
            }
            break;
        default:
            fprintf(stderr,
                    "Usage: %s [-dtpr] [-c address[/port]]\n",
                    argv[0]);
            exit(1);
        }
    }

    if (addr == NULL &&
        ((i = getaddrinfo("127.0.0.1", confd_port, &hints, &addr)) != 0))
        /* "Can't happen" */
        confd_fatal("%s: Failed to get address for ConfD: %s\n",
                    argv[0], gai_strerror(i));
    if ((dname = strrchr(argv[0], '/')) != NULL)
        dname++;
    else
        dname = argv[0];
    /* Init library */
    confd_init(dname, stderr, debuglevel);

    if ((dctx = confd_init_daemon(dname)) == NULL)
        confd_fatal("Failed to initialize ConfD\n");
    if ((ctlsock = get_ctlsock(addr)) < 0)
        confd_fatal("Failed to connect to ConfD\n");
    if ((workersock = get_workersock(addr)) < 0)
        confd_fatal("Failed to connect to ConfD\n");

    memset(&ncb, 0, sizeof(ncb));
    ncb.fd = workersock;
    strcpy(ncb.streamname, "Trader");
    if (confd_register_notification_stream(dctx, &ncb, &live_ctx) != CONFD_OK) {
        confd_fatal("Couldn't register stream %s\n", ncb.streamname);
    }
    if (confd_register_done(dctx) != CONFD_OK) {
        confd_fatal("Failed to complete registration\n");
    }

    struct confd_identityref paperback = {
        bz__ns,
        bz_paperback,
    };
    char* what_we_think_about = "What We Think About When We Try Not To Think About Global Warming: Toward a New Psychology of Climate Action";
    send_shipping_notif("janl", what_we_think_about, paperback, 1);
}
